provider/oracle/environ.go

```diff
@@ -0,0 +1,389 @@
+// cloud only supports amd64.
+func (o *oracleEnviron) Bootstrap(
+	ctx environs.BootstrapContext,
+	params environs.BootstrapParams,
+) (*environs.BootstrapResult, error) {
+	return common.Bootstrap(ctx, o, params)
+}
+
+// BootstrapMessage optionally provides a message to be displayed to
+// the user at bootstrap time.
+func (o oracleEnviron) BootstrapMessage() string {
+	return ""
+}
+
+// Create creates the environment for a new hosted model.
+//
+// This will be called before any workers begin operating on the
+// Environ, to give an Environ a chance to perform operations that
+// are required for further use.
+//
+// Create is not called for the initial controller model; it is
+// the Bootstrap method's job to create the controller model.
+func (o oracleEnviron) Create(params environs.CreateParams) error {
+	return nil
+}
+
+// AdoptResources is called when the model is moved from one
+// controller to another using model migration. Some providers tag
+// instances, disks, and cloud storage with the controller UUID to
+// aid in clean destruction. This method will be called on the
+// environ for the target controller so it can update the
+// controller tags for all of those things. For providers that do
+// not track the controller UUID, a simple method returning nil
+// will suffice. The version number of the source controller is
+// provided for backwards compatibility - if the technique used to
+// tag items changes, the version number can be used to decide how
+// to remove the old tags correctly.
+func (e oracleEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
+
+	return nil
+}
+
+// StartInstance asks for a new instance to be created, associated with
+// the provided config in machineConfig. The given config describes the juju
+// state for the new instance to connect to. The config MachineNonce, which must be
+// unique within an environment, is used by juju to protect against the
+// consequences of multiple instances being started with the same machine
+// id.
+func (o oracleEnviron) StartInstance(args environs.StartInstanceParams) (*environs.StartInstanceResult, error) {
+	if args.ControllerUUID == "" {
+		return nil, errors.NotFoundf("Controller UUID")
+	}
+
+	fmt.Printf("%+v\n", args)
+	os.Exit(1)
+	types, err := getInstanceTypes(o.p.client)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	imagemetadata, err := checkImageList(
+		o.p.client, args.Constraints,
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	//find the best suitable instance returned from the api
+	spec, imagelist, err := findInstanceSpec(
+		o.p.client,
+		imagemetadata,
+		types,
+		&envinstance.InstanceConstraint{
+			Region:      o.spec.Region,
+			Arches:      []string{"amd64"},
+			Constraints: args.Constraints,
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// we are using just the juju ssh keys, no other keys
```

Comment:
is this ssh key uploading really necessary? cloud-init should be taking care of setting up the keys on the new instance